servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -77,91 +80,120 @@ public void onComplete() {
+            assert cItemsPending >= 0;
+            try {
+                state.accumulate(t, target);
+            } finally {
+                if (cItemsPending == 0 && state.requestMore()) {
+                    tSubscription.request(bufferSizeHint);
+                }
             }
         }
 
         @Override
         public void onError(final Throwable t) {
-            state.itemsTerminated(error(t));
+            state.itemsTerminated(error(t), target, bSubscription);
         }
 
         @Override
         public void onComplete() {
-            state.itemsTerminated(complete());
+            state.itemsTerminated(complete(), target, bSubscription);
         }
     }
 
     private static final class BoundariesSubscriber<T, B> implements Subscriber<Accumulator<T, B>> {
         private final State state;
         private final Subscriber<? super B> target;
+        private final DelayedSubscription bSubscription;
         private final Subscription tSubscription;
 
-        @Nullable
-        private ConcurrentSubscription subscription;
-
-        BoundariesSubscriber(final State state, final Subscriber<? super B> target, final Subscription tSubscription) {
+        BoundariesSubscriber(final State state, final Subscriber<? super B> target,
+                             final DelayedSubscription bSubscription, final Subscription tSubscription) {
             this.state = state;
             this.target = target;
+            this.bSubscription = bSubscription;
             this.tSubscription = tSubscription;
         }
 
         @Override
-        public void onSubscribe(final Subscription bSubscription) {
-            subscription = wrap(new Subscription() {
+        public void onSubscribe(final Subscription subscription) {
+            // We may cancel from multiple threads and DelayedSubscription will atomically swap if a cancel occurs but
+            // it will not prevent concurrent access between request(n) and cancel() on the original subscription.
+            bSubscription.delayedSubscription(ConcurrentSubscription.wrap(new Subscription() {
                 @Override
                 public void request(final long n) {
-                    bSubscription.request(n);
+                    subscription.request(n);
                 }
 
                 @Override
                 public void cancel() {
                     try {
-                        bSubscription.cancel();
+                        subscription.cancel();
                     } finally {
                         tSubscription.cancel();
                     }
                 }
+            }));
+            // Wrap bSubscription to count number of requested items excluding internal demand for discarded boundaries.
+            target.onSubscribe(new Subscription() {
+                @Override
+                public void request(final long n) {
+                    final boolean needRequestItems = state.requested(n, target);
```

Comment:
any time there is invalid demand we should propagate it upstream (which is expected to propagate an error back downstream), to avoid accounting for negative values in the `pendingUpdater` state machine you can do something like this:

```java
@Override
                public void request(final long n) {
                    if (SubscriberUtils.isRequestNValid(n)) {
                        final boolean needRequestItems = state.requested(n, target);
                        bSubscription.request(n);
                        if (needRequestItems) {
                            tSubscription.request(state.itemsRequestN);
                        }
                    } else {
                        try {
                            bSubscription.request(n);
                        } finally {
                            tSubscription.request(n);
                        }
                    }
                }
```